# Datasets
Sequencing data used in this repository are from the Genome in a Bottle (GIAB) project, specifically the Ashkenazim Trio son (HG002).
- **GIAB**: [https://ftp.ncbi.nlm.nih.gov/ReferenceSamples/giab/data/AshkenazimTrio/HG002_NA24385_son/](https://ftp.ncbi.nlm.nih.gov/ReferenceSamples/giab/data/AshkenazimTrio/HG002_NA24385_son/)
<br><br>

Two sequencing datasets were downloaded from GIAB: one short-read (Ilumina) and one long-read dataset (ONT Ultra-long).

| Read | Dataset | File Name | File Size | 
|:----:|:-------:|:---------:|:---------:|
|  Short-read (SR)  |  NIST_Illumina_2x250bps  |  D1_S1_L001_R1_001.fastq.gz  | 5.6G |
|  Long-read (LR)  |  UCSC_Ultralong_OxfordNanopore_Promethion  |  GM24385_1.fastq.gz  | 27G | 
<br>

Because the dataset was too large to iterate, I extracted a subset for benchmarking.

```
# Sampling LR dataset with sample fraction 0.05
seqtk_rs sample -I ${FASTQ} --random-seed 11 -f 0.05 | gzip > ${SUBSET_FQ}

# Sampling SR dataset with sample fraction 0.2
seqtk_rs sample -I ${FASTQ} --random-seed 11 -f 0.2 | gzip > ${SUBSET_FQ}

# Calculate FASTA/FASTQ statistics
# Output: #seq, #bases, avg_size, min_size, med_size, max_size, N50
seqtk_rs size -A ${FASTA}
seqtk_rs size -I ${FASTQ}
```
<br>


FASTA files was generated by `seqtk_rs`:

```
seqtk_rs seq -I ${FASTQ} --output-fasta | gzip > ${FASTA}
```
<br>

The stats of subsets:
| Read | Subset Name | File Size | #Sequences | #Bases | Avg. Size | Min. Size | Median Size | Max Size | N50 |
|:----:|:-----------:|:---------:|:----------:|:------:|:---------:|:---------:|:-----------:|:--------:|:---:|
|  SR  | D1_S1_L001_R1_001_subset.fastq.gz | 1.1G | 6,785,633 | 1,684,541,053 | 248.25 | 35 | 250 | 250 | 250 |
|  LR  | GM24385_1_subset.fastq.gz | 1.3G | 145,025 | 1,325,261,486 | 9,138.16 | 5 | 1,043 | 290,123 | 50,242 |
<br> 

[Back to ReadMe](../readme.md)
